File: python/sasktran2/engine.py
import sasktran2 as sk


class Engine(object):
    def __init__(self, config: sk.Config, model_geometry: sk.Geometry1D, viewing_geo: sk.ViewingGeometry, nstokes: int = 1
                 ):
        self._nstokes = nstokes

        if nstokes == 1:
            self._engine = sk.EngineStokes_1(config, model_geometry, viewing_geo)
        elif nstokes == 3:
            self._engine = sk.EngineStokes_3(config, model_geometry, viewing_geo)

    def calculate_radiance(self, atmosphere: sk.Atmosphere, output: sk.Output = None):
        if output is None:
            engine_output = sk.OutputIdeal(self._nstokes)
        else:
            engine_output = output

        self._engine.calculate_radiance(atmosphere._atmosphere, engine_output.internal_output())

        return engine_output.post_process()

File: src/engine.cpp
#include <pybind11/pybind11.h>
#include <pybind11/eigen.h>
#include <sasktran2.h>

namespace py = pybind11;
using namespace pybind11::literals;

template <int NSTOKES>
void declareEngine(py::module_& m, const std::string& suffix) {
    using Engine = Sasktran2<NSTOKES>;

    py::class_<Engine>(m, ("Engine" + suffix).c_str())
        .def(py::init<
                 const sasktran2::Config&, const sasktran2::Geometry1D*,
                 const sasktran2::viewinggeometry::ViewingGeometryContainer&>(),
             R"(
                 Internal SASKTRAN2 object which handles the radiative transfer calculation.

                 Parameters
                 ----------
                 config: sasktran2.Config
                     Configuration settings
                 model_geometry: sasktran2.Geometry1D
                     The model geometry
                 viewing_geometry: sasktan2.ViewingGeometry
                     The viewing geometry
                 )",
             "config"_a, "model_geometry"_a, "viewing_geometry"_a)
        .def("calculate_radiance", &Engine::calculate_radiance,
             R"(
                    Performs the radiative transfer calculation for the given atmosphere, placing the result in output

                    Parameters
                    ----------
                    atmosphere: sasktran2.Atmosphere
                        The atmosphere object

                    output: sasktran2.Output
                        The result to place the output inside
                 )",
             "atmosphere"_a, "output"_a);
}

void init_engine(py::module_& m) {
    declareEngine<1>(m, "Stokes_1");
    declareEngine<3>(m, "Stokes_3");
}
